### DISCUSSION TITLE
How can I stack a window on top of the one further on the right?

### DISCUSSION THREAD

**Discussion by stefanopagliari** [asmvik/yabai#1312] [Q&A]
I have a widescreen monitor and I usually stack on the right end of the monitor apps that are not the main activity I'm working on the moment (e.g. email, calendar, etc...).
Is there a way to set a key so that a newly opened app is stacked upon the app on the right?

**Comment by stefanopagliari** [comment]
I ended up using this but I'm sure some more elegant solutions exist:

```
shift + cmd + ctrl + alt - l : yabai -m window --swap next;\
yabai -m window --swap next;\
yabai -m window --swap next;\
yabai -m window --focus prev;\
yabai -m window --stack next
```

**Reply by diocletiann** [reply]
https://github.com/koekeishiya/yabai/discussions/1226#discussioncomment-3006929

You can change the insert/warp commands under `n > 3` commands to stack into `sp.LastWindow` if you want.

**Comment by diocletiann** [comment]
Here you go, I added two focus commands to fix the ID glitch that happens when new stack is created:

```go
package main

import (
    "encoding/json"
    "io"
    "net"
    "os"
    "os/user"
    "strconv"
    "strings"
)

type window struct {
    ID        int    `json:"id"`
    IsVisible bool   `json:"is-visible"`
    SplitType string `json:"split-type"`
}

type space struct {
    LastWindow int `json:"last-window"`
}

func main() {
    newId := os.Args[1]
    newIdInt, _ := strconv.Atoi(newId)

    u, _ := user.Current()
    addr := "/tmp/yabai_" + u.Username + ".socket"

    var w []window
    c, _ := net.Dial("unix", addr)
    c.Write([]byte("query\x00--windows\x00--space\x00\x00"))
    r, _ := io.ReadAll(c)
    json.Unmarshal(r, &w)

    var n int
    for _, v := range w {
        if v.IsVisible {
            n++
        }
    }
    if n <= 3 {
        for _, v := range w {
            if v.ID == newIdInt && v.SplitType == "horizontal" {
                s := strings.Join([]string{"window", newId, "--toggle", "split", "\x00"}, "\x00")
                c, _ = net.Dial("unix", addr)
                c.Write([]byte(s))

                s = strings.Join([]string{"space", "--balance", "\x00"}, "\x00")
		c, _ = net.Dial("unix", addr)
		_, _ = c.Write([]byte(s))
                return
            } else {
		s := strings.Join([]string{"space", "--balance", "\x00"}, "\x00")
		c, _ = net.Dial("unix", addr)
		_, _ = c.Write([]byte(s))
                return
	    }
        }
    }
    if n > 3 {
        var sp space
        c, _ = net.Dial("unix", addr)
        c.Write([]byte("query\x00--spaces\x00--space\x00\x00"))
        r, _ = io.ReadAll(c)
        json.Unmarshal(r, &sp)
        targetWindow := strconv.Itoa(sp.LastWindow)
        
        if newIdInt == sp.LastWindow {
            targetWindow = "prev"
        }

        c, _ = net.Dial("unix", addr)
        s := strings.Join([]string{"window", targetWindow, "--stack", newId, "\x00"}, "\x00")
        c.Write([]byte(s))

        c, _ = net.Dial("unix", addr)
        s = strings.Join([]string{"window", "--focus", "stack.prev", "\x00"}, "\x00")
        c.Write([]byte(s))

        c, _ = net.Dial("unix", addr)
        s = strings.Join([]string{"window", "--focus", "stack.next", "\x00"}, "\x00")
        c.Write([]byte(s))
    }
}

**Reply by stefanopagliari** [reply]
@restfuladi Thank you for this.
Should I put this in a sh file and add to my .yabairc something like
yabai -m signal --add event=window_created action='PATHTOFILE.sh $YABAI_WINDOW_ID'

**Reply by diocletiann** [reply]
You're welcome. Paste the code into a .go file and compile with `go build filename.go`. Then invoke the new binary like in your example but without the `.sh`.

**Reply by stefanopagliari** [reply]
> 

Thanks. Sorry,  I'm not familiar with go. 

I get the error `no new variables on left side of :=` at line 49

Line 49 is
`s := strings.Join([]string{"space", newId, "--balance", "\x00"}, "\x00")`

**Reply by diocletiann** [reply]
My fault, I just noticed that too. Delete the `:` and it should compile. I fixed the code above.

**Reply by diocletiann** [reply]
found a bug, stacking into `last-window` only works when it's not focused. I'll look into a solution.

**Reply by diocletiann** [reply]
Ok the latest edit seems to work as intended, make sure you paste the entire text, I found another bug.

**Reply by stefanopagliari** [reply]
@restfuladi Thank you for this! Much appreciated

**Reply by diocletiann** [reply]
YW. You can also enable auto-balance in yabairc and recompile without the balance commands. It's easier IMO and also works when closing windows.
